rewrite-maven/src/main/java/org/openrewrite/maven/AddManagedDependency.java

```diff
@@ -0,0 +1,173 @@
+                    "Setting 'version' to \"25-29\" can be paired with a metadata pattern of \"-jre\" to select 29.0-jre",
+            example = "-jre",
+            required = false)
+    @Nullable
+    private String versionPattern;
+
+    @Option(displayName = "Releases only",
+            description = "Whether to exclude snapshots from consideration when using a semver selector",
+            example = "true",
+            required = false)
+    @Nullable
+    private Boolean releasesOnly;
+
+    @Option(displayName = "Only if using glob expression for group:artifact",
+            description = "Only add managed dependencies to projects having a concrete dependency matching the expression.",
+            example = "org.apache.logging.log4j:log4j*")
+    @Nullable
+    private String onlyIfUsing;
+
+    @Option(displayName = "Add to the root pom",
+            description = "Add to the root pom where root is the eldest parent of the pom within the source set.",
+            example = "true",
+            required = false)
+    private Boolean addToRootPom;
+
+    @Override
+    public Validated validate() {
+        Validated validated = super.validate();
+        //noinspection ConstantConditions
+        if (version != null) {
+            validated = validated.or(Semver.validate(version, versionPattern));
+        }
+        return validated;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Add Managed Maven Dependency";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        if (!StringUtils.isNullOrEmpty(onlyIfUsing) && onlyIfUsing.contains(":")) {
+            String[] ga = onlyIfUsing.split(":");
+            return new MavenIsoVisitor<ExecutionContext>() {
+                @Override
+                public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext context) {
+                    Xml.Tag t = super.visitTag(tag, context);
+
+                    if (isDependencyTag()) {
+                        ResolvedDependency dependency = findDependency(t, null);
+                        if (dependency != null) {
+                            ResolvedDependency match = dependency.findDependency(ga[0], ga[1]);
+                            if (match != null) {
+                                t = t.withMarkers(t.getMarkers().searchResult());
+                            }
+                        }
+                    }
+
+                    return t;
+                }
+            };
+        }
+        return null;
+    }
+
+    @Override
+    protected List<SourceFile> visit(List<SourceFile> before, ExecutionContext ctx) {
+        List<SourceFile> rootPoms = new ArrayList<>();
+        for (SourceFile source : before) {
+            source.getMarkers().findFirst(MavenResolutionResult.class).ifPresent(mavenResolutionResult -> {
+                if (mavenResolutionResult.getParent() == null) {
+                    rootPoms.add(source);
```

Comment:
what does this look like when a pom has a parent that is external to the project? 

don't we need something like

```
(mavenResolutionResult.getParent == null || mavenResolutionResult.getParent().getPom().getRequested().getSourcePath() == null)
```

Reply:
it works check out the `Added to the root pom having parent not within the source set` test
https://github.com/openrewrite/rewrite/pull/1436/files#diff-bc328d6a020c9de3a4b64464ecb3daf052dd19d5e0bba9418716ba901a198d80R235

Reply:
👍